### DISCUSSION TITLE
Awaiting ready status of models

### DISCUSSION THREAD

**Discussion by zekth** [dynamoose/dynamoose#1218] [General] [answered]
While creating Models with options `create` and `waitForActive` it seems there is no proper way to await the readiness of those before continuing. I tried to wait for `pendingTaskPromise `: https://github.com/dynamoose/dynamoose/blob/aef6de3cf2a5aa7d3400174115240a687b8cef8f/lib/Model/index.ts#L307
But it seems to resolve before the connection was properly established.

I'd do something
```js
  const users = dynamoose.model(
    'Users',
    new dynamoose.Schema(
      {
        ....
      },
      {
        timestamps: true
      }
    ),
    { waitForActive: { enabled: true } }
  );
await users.readyToUse(); //
// do something
```

Is something like this possible in the current state?

**Comment by fishcharlie** [comment, marked as answer]
@zekth Anytime you take an action that relies on the model to be ready (ie. creating a document, running a query/scan, retrieving a document, etc), it will automatically wait for the model to be ready to use.

Is there another use case you are thinking about here?

**Reply by zekth** [reply]
I mean i want to wait for the models to be ready before starting the server. So one possibility is to send a query in each model to ensure all models are ready before starting the server?

**Reply by fishcharlie** [reply]
@zekth 

> i want to wait for the models to be ready before starting the server

Why tho? Like what is the use case here? In development, there isn't any problem with the server being a little bit slower to make those initial database comments. In production, you shouldn't even be setting `create` or `waitForActive` to true. Both of those should be set to false in production and the table(s) should already be setup and ready to go.

> So one possibility is to send a query in each model to ensure all models are ready before starting the server?

I mean sure? There is a cost and performance impact for that. It'd achieve what you want, but it doesn't take into account the negatives of it.

---

If you want to give more information about your use case and want to submit a feature request, feel free to create an issue for that. But as it stands right now, I don't see the use case or purpose for this, so I don't think it'd get implemented. If you believe this is necessary tho and would be beneficial, I'd love to hear more about why so we can consider implementing it.

**Reply by zekth** [reply]
Sure your explanation makes sense.

In my use case i want the server to return 503 http until all direct dependencies are available. For example doing a rolling update with K8S (like using `helm upgrade --atomic`) the deployment will fail if it has been setup with some wrong environment values like credentials and such. It might be over-engineered in some context but it ensures you don't upgrade in a bad shape.

In one other case, you have your DB set with `IaC` but you haven't merged the last changes in the target environment but code has already been merged for the service. Service starts but when it tries to query the new table it fails but it's still deployed.

Sure it has cost impact on the cold start of your service / function, but only at startup not at runtime.

If you think this kind of use case is valuable i'd work on it for sure.

**Reply by fishcharlie** [reply]
@zekth As of https://github.com/dynamoose/dynamoose/pull/1317, in v3, you'll be able to call `table.initialize` to setup your table, which returns a promise or takes a callback function. You can combine this with the `initialize: false` option in your table settings, which should help you achieve what you want.

**Reply by zekth** [reply]
Thanks!
